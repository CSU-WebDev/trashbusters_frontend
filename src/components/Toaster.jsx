import React, { useEffect, useState } from 'react';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Toast from 'react-bootstrap/Toast';

export default function Toaster(props) {
    const [show, setShow] = useState(true);
    useEffect(() => {
        setShow( props.apiStatus === "ERROR" ? true : false)
    },[props.apiStatus])
    if (props.apiStatus === 'ERROR') {
        return (
        <Row data-testid='toaster-container'>
            <Col md={6} className="mb-2">
            <Toast 
                data-testid='toaster'
                show={show} 
                onClose={() => {
                    props.setApiStatus(null) 
                    setShow(false)}
                }>
                <Toast.Header>
                <img
                    src="holder.js/20x20?text=%20"
                    className="rounded me-2"
                    alt=""
                />
                <strong className="me-auto">Error</strong>
                </Toast.Header>
                <Toast.Body>{props.apiMessage}</Toast.Body>
            </Toast>
            </Col>
        </Row>
        )
    }
    return null

}