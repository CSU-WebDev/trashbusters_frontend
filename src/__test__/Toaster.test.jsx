import React from 'react';
import { render, screen, fireEvent, prettyDOM } from '@testing-library/react';
import Toaster from '../components/Toaster';
import '@testing-library/jest-dom';

describe('Toaster component', () => {
  test('matches snapshot when there is no error', () => {
    const { asFragment } = render(<Toaster apiStatus="SUCCESS" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('matches snapshot when there is an error', () => {
    const { asFragment } = render(<Toaster apiStatus="ERROR" apiMessage="Something went wrong" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders Toaster component when there is an error', () => {
    render(<Toaster apiStatus="ERROR" apiMessage="Something went wrong" />);
    const toasterElement = screen.getByTestId('toaster');
    expect(toasterElement).toBeInTheDocument();
  });

  test('does not render Toaster component when there is no error', () => {
    render(<Toaster apiStatus="SUCCESS" />);
    const toasterElement = screen.queryByTestId('toaster');
    expect(toasterElement).not.toBeInTheDocument();
  });

  test('renders error message when there is an error', () => {
    render(<Toaster apiStatus="ERROR" apiMessage="Something went wrong" />);
    const toasterBody = screen.getByText(/Something went wrong/i);
    expect(toasterBody).toBeInTheDocument();
  });

  test('renders correct error header', () => {
    render(<Toaster apiStatus="ERROR" apiMessage="Something went wrong" />);
    const toasterHeader = screen.getByText(/Error/i);
    expect(toasterHeader).toBeInTheDocument();
  });

  // test('calls setApiStatus with null and hides the Toast when the close button is clicked', () => {
  //   const setApiStatusMock = jest.fn((value) => {
  //     console.log('setApiStatusMock called with', value);
  //   });      
  //   render(<Toaster apiStatus="ERROR" apiMessage="Something went wrong" setApiStatus={setApiStatusMock} />);

  //   // Find the close button and click it
  //   const closeButton = screen.getByRole('button');
  //   fireEvent.click(closeButton);

  //   // Verify that setApiStatus was called with null
  //   expect(setApiStatusMock).toHaveBeenCalledWith(null);

  //   // Verify that the Toast is not visible anymore
  //   const toasterElement = screen.queryByTestId('toaster');
  //   console.log(prettyDOM(toasterElement));
  //   expect(toasterElement).toBeNull();
  // });
  
});
