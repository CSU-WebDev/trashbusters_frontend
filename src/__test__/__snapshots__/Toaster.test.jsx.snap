// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster component matches snapshot when there is an error 1`] = `
<DocumentFragment>
  <div
    class="row"
    data-testid="toaster-container"
  >
    <div
      class="mb-2 col-md-6"
    >
      <div
        aria-atomic="true"
        aria-live="assertive"
        class="fade toast show"
        data-testid="toaster"
        role="alert"
      >
        <div
          class="toast-header"
        >
          <img
            alt=""
            class="rounded me-2"
            src="holder.js/20x20?text=%20"
          />
          <strong
            class="me-auto"
          >
            Error
          </strong>
          <button
            aria-label="Close"
            class="btn-close"
            data-dismiss="toast"
            type="button"
          />
        </div>
        <div
          class="toast-body"
        >
          Something went wrong
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toaster component matches snapshot when there is no error 1`] = `<DocumentFragment />`;
